### PR TITLE
Swift 3.0 Migration

### LINKED ISSUE #240: Swift 3.0 support - bitcode support
I created a `Swift3` branch on my fork and am working on Swift 3.0 support.

### PR DESCRIPTION
**Pull request by colemancda**
Cleaner version of #241. Fixes #240.

### CODE REVIEW

File: Sources/CryptoSwift/LinuxFoundation.swift [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+#if !os(Linux)
+    
+extension String {
+    
+    // Workaround:
+    // https://github.com/krzyzanowskim/CryptoSwift/issues/177
+    @inline(__always)
+    func bridge() -> NSString {
+        return self as NSString
```
Comment (krzyzanowskim) [inline Sources/CryptoSwift/LinuxFoundation.swift]: this workaround was added for Linux only. Now if it's not needed (I assume it's not because this code is for OSX only) no need to have it any longer.

File: Sources/CryptoSwift/LinuxFoundation.swift [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    }
+}
+
+/// Fixes discrepancies with Swift 3.0 Linux
```
Comment (krzyzanowskim) [inline Sources/CryptoSwift/LinuxFoundation.swift]: can you please elaborate a little about why do we have to add custom Base64 implementation?

File: Sources/CryptoSwift/LinuxFoundation.swift [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        
+        func append(_ bytes: UnsafePointer<Void>, length: Int) {
+            
+            appendBytes(bytes, length: length)
+        }
+    }
+    
+    public extension NSDate {
+        
+        func timeIntervalSince(_ date: NSDate) -> NSTimeInterval {
```
Comment (krzyzanowskim) [inline Sources/CryptoSwift/LinuxFoundation.swift]: Is this required for CryptoSwift ?

File: Sources/CryptoSwift/IntegerConvertible.swift [hunk 4]
```diff
@@ -9,8 +9,8 @@
 protocol BitshiftOperationsType {
     func <<(lhs: Self, rhs: Self) -> Self
     func >>(lhs: Self, rhs: Self) -> Self
-    func <<=(inout lhs: Self, rhs: Self)
-    func >>=(inout lhs: Self, rhs: Self)
+    func <<=( lhs: inout Self, rhs: Self)
+    func >>=( lhs: inout Self, rhs: Self)
```
Comment (krzyzanowskim) [inline Sources/CryptoSwift/IntegerConvertible.swift]: no spaces please ;)))

File: Sources/CryptoSwift/LinuxFoundation.swift [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+            getBytes(&decodedBytes, length: decodedBytes.count)
+            let encodedBytes = NSData.base64EncodeBytes(decodedBytes, options: options)
+            let characters = encodedBytes.map { Character(UnicodeScalar($0)) }
+            return String(characters)
+        }
+    }
+    
+    public extension NSMutableData {
+        
+        func append(_ bytes: UnsafePointer<Void>, length: Int) {
```
Comment (krzyzanowskim) [inline Sources/CryptoSwift/LinuxFoundation.swift]: why do we need this?

File: Sources/CryptoSwift/Multiplatform.swift [hunk 6]
```diff
@@ -13,10 +13,10 @@
     import Darwin
 #endif
 
-func cs_arc4random_uniform(upperBound: UInt32) -> UInt32 {
+func cs_arc4random_uniform(_ upperBound: UInt32) -> UInt32 {
     #if os(Linux)
-        return _swift_stdlib_arc4random_uniform(upperBound)
+        return _swift_stdlib_cxx11_mt19937_uniform(upperBound)
```
Comment (krzyzanowskim) [inline Sources/CryptoSwift/Multiplatform.swift]: Note: caused by https://github.com/apple/swift/pull/1939 and arc4random() is removed now completely.

File: Sources/CryptoSwift/PBKDF2.swift [hunk 7]
```diff
@@ -6,12 +6,16 @@
 //  Copyright © 2016 Marcin Krzyzanowski. All rights reserved.
 //
 
+#if os(Linux)
+    import Glibc
```
Comment (krzyzanowskim) [inline Sources/CryptoSwift/PBKDF2.swift]: shouln't be needed here

File: Sources/CryptoSwift/CipherType.swift [hunk 8]
```diff
@@ -12,20 +12,20 @@
...
-public enum CipherError: ErrorType {
+public enum CipherError: ErrorProtocol {
     case Encrypt
     case Decrypt
 }
 
-public protocol CipherType {
-    func cipherEncrypt(bytes: [UInt8]) throws -> [UInt8]
-    func cipherDecrypt(bytes: [UInt8]) throws -> [UInt8]
+public protocol Cipher {
```
Comment (krzyzanowskim) [inline Sources/CryptoSwift/CipherType.swift]: ? why you change CipherType back to Cipher?

File: Sources/CryptoSwiftTests/AESTests.swift [hunk 9]
```diff
@@ -5,16 +5,161 @@
...
+    
     // 128 bit key
     let aesKey:[UInt8] = [0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f]
+    
+    // MARK: - Functional Tests
 
-    func testAES_encrypt2() {
-        let key:[UInt8]   = [0x36, 0x37, 0x39, 0x66, 0x62, 0x31, 0x64, 0x64, 0x66, 0x37, 0x64, 0x38, 0x31, 0x62, 0x65, 0x65];
-        let iv:[UInt8]    = [0x6b, 0x64, 0x66, 0x36, 0x37, 0x33, 0x39, 0x38, 0x44, 0x46, 0x37, 0x33, 0x38, 0x33, 0x66, 0x64]
+    func testAES_encrypt4() {
```
Comment (krzyzanowskim) [inline Sources/CryptoSwiftTests/AESTests.swift]: Would be good have this change as separate PR with more description. Especially why this test looks like this.

File: Sources/CryptoSwiftTests/ExtensionsTest.swift [hunk 10]
```diff
@@ -100,5 +87,19 @@ final class ExtensionsTest: XCTestCase {
         let decrypted = try! encryptedBase64.decryptBase64ToString(AES(key: "secret0key000000", iv: "0123456789012345"))
         XCTAssertEqual(decrypted, "my secret string")
     }
-
+    
+    // MARK: - Performance Tests
+    
+    #if !os(Linux)
```
Comment (krzyzanowskim) [inline Sources/CryptoSwiftTests/ExtensionsTest.swift]: this is worth a comment why we disable perf tests on Linux

**Comment by krzyzanowskim** [comment]
@colemancda A lot of awesome work here, thanks! few changes may need more comments thought, Please review my comments.

File: Package.swift [hunk 11]
```diff
@@ -1,5 +1,12 @@
 import PackageDescription
 
 let package = Package(
-    name: "CryptoSwift"
+    name: "CryptoSwift",
+    targets: [
+                Target(
+                    name: "CryptoSwiftTests",
+                    dependencies: [.Target(name: "CryptoSwift")]),
```
Comment (czechboy0) [inline Package.swift]: This is not needed, SwiftPM automatically creates a Tests target when there's a folder called `Tests` with `CryptoSwiftTests` inside of it.

File: Sources/CryptoSwiftTests/ExtensionsTest.swift (hunk 10 above)
Comment (czechboy0) [inline Sources/CryptoSwiftTests/ExtensionsTest.swift]: XCTest on Linux doesn't yet have performance testing.

**Comment by czechboy0** [comment]
👍 Can't wait for this to get merged.

File: Sources/CryptoSwift/LinuxFoundation.swift (hunk 2 above)
Comment (colemancda) [inline Sources/CryptoSwift/LinuxFoundation.swift]: The open source version of Foundation is having problems importing into Swift 3.0, this is only a temporary work around. The compiler said that the base64 methods did not exist on Linux.

File: Sources/CryptoSwift/LinuxFoundation.swift (hunk 3 above)
Comment (colemancda) [inline Sources/CryptoSwift/LinuxFoundation.swift]: Yes, there are differences between in the way Swift 3.0 imports Cocoa APIs, and the Open Source version of Foundation that still has the Swift 2.2 syntax.

File: Sources/CryptoSwift/CipherType.swift (hunk 8 above)
Comment (colemancda) [inline Sources/CryptoSwift/CipherType.swift]: To follow the Swift 3.0 API guidelines

https://github.com/apple/swift-evolution/blob/master/proposals/0006-apply-api-guidelines-to-the-standard-library.md

File: Sources/CryptoSwiftTests/AESTests.swift (hunk 9 above)
Comment (colemancda) [inline Sources/CryptoSwiftTests/AESTests.swift]: AES encrypting with 512 bit key crashes on Linux. testAES_encrypt4() is code from a project I'm working on.

File: Package.swift (hunk 11 above)
Comment (colemancda) [inline Package.swift]: True, in theory... That feature of SwiftPM crashes when you try to run it. :(

**Comment by colemancda** [comment]
@czechboy0 @krzyzanowskim Pls check my responses to your comments, hopefully I've addressed all of them.

File: Sources/CryptoSwift/LinuxFoundation.swift (hunk 2 above)
Comment (krzyzanowskim) [inline Sources/CryptoSwift/LinuxFoundation.swift]: I see. In that case, better disable base64 related code on Linux than copy and maintain foundation. It's temporary after all.

File: Sources/CryptoSwift/LinuxFoundation.swift (hunk 3 above)
Comment (krzyzanowskim) [inline Sources/CryptoSwift/LinuxFoundation.swift]: ah, this should be removed completly (my fault).

File: Sources/CryptoSwift/LinuxFoundation.swift (hunk 5 above)
Comment (krzyzanowskim) [inline Sources/CryptoSwift/LinuxFoundation.swift]: @colemancda

File: Sources/CryptoSwift/NSData+Extension.swift [hunk 12]
```diff
@@ -32,78 +31,78 @@ extension NSData {
...
         let result = try authenticator.authenticate(self.arrayOfBytes())
-        return NSData.withBytes(result)
+        return NSData.with(bytes: result)
     }
 }
 
 extension NSData {
     
-    public func toHexString() -> String {
+    final public func toHexString() -> String {
```
Comment (krzyzanowskim) [inline Sources/CryptoSwift/NSData+Extension.swift]: what about this final, it is required somewhere now ?

File: Sources/CryptoSwiftTests/AESTests.swift (hunk 9 above)
Comment (krzyzanowskim) [inline Sources/CryptoSwiftTests/AESTests.swift]: I see, this function looks odd anyway. Do you think you can make it any easier to follow? like, few lines long similar to rest of the tests?

File: Sources/CryptoSwift/CipherType.swift (hunk 8 above)
Comment (krzyzanowskim) [inline Sources/CryptoSwift/CipherType.swift]: I prefer CipherProtocol. I just removed Cipher (enum) and this may lead to more confusion.

**Comment by krzyzanowskim** [comment]
Ok. @colemancda do you think you can split changes in few PR so I can merge in parts? I don't want merge it all at once as I'm not sure about the whole. 
PR is great, it just feels like too much changes at once.

File: Sources/CryptoSwiftTests/AESTests.swift (hunk 9 above)
Comment (krzyzanowskim) [inline Sources/CryptoSwiftTests/AESTests.swift]: btw... there is no AES-512. Anyway... can't find reason for this test in this form atm ;)

**Comment by krzyzanowskim** [comment]
I did another look at changes today.

@colemancda Great piece of work here, but I won't merge PR in this form right now.

My suggestions::
- split changes to several PR's
- don't reorganise project where not necessary (at least not at once)
- remove test (or replace with something that have more sense)
- disable Base64 part for Linux at all and mark as workaround for Linux Foundation limitations.
